micro/walk.go

```diff
@@ -40,11 +40,12 @@ func assv(v *ast.Variable, ss Substitutions) (*ast.SExpr, bool) {
 	if ss == nil {
 		return nil, false
 	}
-	value, ok := ss[v.Index]
-	if !ok {
-		return nil, false
+	for _, pair := range ss {
```

Comment:
So this should probably result in things becoming slower.
Is there no way to still do a lookup, instead of looping through all the pairs one by one.
Taking us to O(n) instead or O(1) ?

Reply:
The tradeoff here is between looping through a list for lookups vs copying maps on each update to state. Far as I can tell, looping is a lot faster.

Reply:
See https://github.com/awalterschulze/gominikanren/pull/19/files#diff-a40591217a88cf468e387aeab3ce5bbcc85280224972538fc906cc5132395665L57

Reply:
Ah okay, but here we are still copying over the slice

```
m := make(Substitutions, len(s)+1)
		copy(m, s)
		m[len(s)] = SubPair{Key: vv.Atom.Var.Index, Value: n}
```

Reply:
Just reverted to the map copy and ran tests to compare execution times, here's the result:
map lookup: % go test ./...
ok  	github.com/awalterschulze/gominikanren/example/einstein	1.210s
ok  	github.com/awalterschulze/gominikanren/example/fibonacci	4.525s
ok  	github.com/awalterschulze/gominikanren/example/fivesandsixes	0.649s
ok  	github.com/awalterschulze/gominikanren/example/mergesort	0.771s
ok  	github.com/awalterschulze/gominikanren/example/peano	0.845s

substitution list: % go test ./...
ok  	github.com/awalterschulze/gominikanren/example/einstein	0.790s
ok  	github.com/awalterschulze/gominikanren/example/fibonacci	1.626s
ok  	github.com/awalterschulze/gominikanren/example/fivesandsixes	0.159s
ok  	github.com/awalterschulze/gominikanren/example/mergesort	0.599s
ok  	github.com/awalterschulze/gominikanren/example/peano	0.231s

Reply:
Wow big improvement, well that sells it :D